### ISSUE TITLE
[Spacy Nightly] init fill-config on textcat transformer produces pooling error

### ISSUE THREAD

**Issue by thomashacker** [explosion/spaCy#6401]
## How to reproduce the behaviour

Hey, i am trying to get a training config for a transformer-based textcat pipeline.

The fix [#6371](https://github.com/explosion/spaCy/pull/6371) seems not to be working for me :( 
Using `init fill-config` on the transformer textcat config (CLI) still produced the pooling error:

✘ Can't construct config: calling registry function
(transformer_listener_tok2vec_v1) failed
spacy-transformers.TransformerListener.v1   transformer_listener_tok2vec_v1() missing 1 required positional argument: 'pooling'

---------------------------------------------------------------------------------------

I tried fixing this by manually adding the pooling attribute but then this error occurs:

✘ Can't construct config: calling registry function
(build_text_classifier) failed
spacy.TextCatEnsemble.v2   Cannot get dimension 'nO' for model 'transformer-listener>>trfs2arrays': value unset

{'model': {'@architectures': 'spacy.TextCatEnsemble.v2', 'nO': None, 'linear_model': {'@architectures': 'spacy.TextCatBOW.v1', 'exclusive_classes': False, 'ngram_size': 1, 'no_output_layer': False}, 'tok2vec': {'@architectures': 'spacy-transformers.TransformerListener.v1', 'grad_factor': 1.0, 'pooling': {'@layers': 'reduce_mean.v1'}}}}

---------------------------------------------------------------------------------------

## Info about spaCy
- **spaCy version:** 3.0.0rc2
- **Platform:** Linux-5.4.0-52-generic-x86_64-with-glibc2.27
- **Python version:** 3.8.0
- **Pipelines:** en_core_web_trf (3.0.0a0),  en_core_web_lg (3.0.0a0),

**Comment by svlandeg** [comment]
Do you have the 3.0.0rc2 installed with `pip`, or did you compile from source? 

The fixes in PR #6371 haven't made it into a release yet - they'll be available from 3.0.0rc3 onwards. If you don't want to wait, you can either install the `develop` branch from source, or make the corresponding edits as you can see them in https://github.com/explosion/spaCy/pull/6371/files. 

Let us know if that works for you?

**Comment by thomashacker** [comment]
I installed via pip. Tried fixing by making the edits and `init fill-config` works now!
But now the `train` produces this error:

Data validation error in 'para-attn'
X: <class 'thinc.types.Ragged'> Y: <class 'numpy.ndarray'>

Y -> 0   instance of Ragged expected

I will wait for the next fix 👍 
Thank you very much!

**Comment by svlandeg** [comment]
Hm, I don't think I've seen that error before. Can you give me the full command(s) you ran?

**Comment by ahalterman** [comment]
I'm getting the same error from spaCy develop, compiled from source (GIT_VERSION = 96726ec1f)

```
python -m spacy debug data config_test.cfg --paths.train crowd_train_small.spacy --paths.dev crowd_train_small.spacy

============================ Data file validation ============================
✔ Corpus is loadable
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/root/spaCy/spacy/__main__.py", line 4, in <module>
    setup_cli()
  File "/root/spaCy/spacy/cli/_util.py", line 65, in setup_cli
    command(prog_name=COMMAND)
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/opt/conda/lib/python3.8/site-packages/typer/main.py", line 497, in wrapper
    return callback(**use_params)  # type: ignore
  File "/root/spaCy/spacy/cli/debug_data.py", line 62, in debug_data_cli
    debug_data(
  File "/root/spaCy/spacy/cli/debug_data.py", line 105, in debug_data
    nlp.initialize(lambda: train_dataset)
  File "/root/spaCy/spacy/language.py", line 1232, in initialize
    proc.initialize(get_examples, nlp=self, **p_settings)
  File "/root/spaCy/spacy/pipeline/textcat.py", line 365, in initialize
    self.model.initialize(X=doc_sample, Y=label_sample)
  File "/opt/conda/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/layers/concatenate.py", line 129, in init
    layer.initialize(X=X, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/model.py", line 294, in initialize
    validate_fwd_input_output(self.name, self._func, X, Y)
  File "/opt/conda/lib/python3.8/site-packages/thinc/util.py", line 424, in validate_fwd_input_output
    raise DataValidationError(name, X, Y, e.errors()) from None
thinc.util.DataValidationError:

Data validation error in 'para-attn'
X: <class 'thinc.types.Ragged'> Y: <class 'numpy.ndarray'>

Y -> 0   instance of Ragged expected
```

Here's the config file and data:

[crowd_train_small.spacy.zip](https://github.com/explosion/spaCy/files/5560056/crowd_train_small.spacy.zip)
[config_test.cfg.zip](https://github.com/explosion/spaCy/files/5560057/config_test.cfg.zip)

And I don't think it should matter, but this is running in a Docker container for CUDA reasons.

**Comment by thomashacker** [comment]
After the edits i could run `spacy init fill-config cls_config.cfg cls_config_full.cfg` without an error:
`
✔ Auto-filled config with all values
✔ Saved config
cls_config_full.cfg
`


Then i tried to train the pipeline with  `spacy train cls_config_full.cfg --output cls/transformer  --paths.train data/cls_data/train.spacy --paths.dev data/cls_data/test.spacy`


This is the full console log after the command:

`ℹ Using CPU

=========================== Initializing pipeline ===========================
Set up nlp object from config
Pipeline: ['transformer', 'textcat']
Created vocabulary
Finished initializing nlp object
Traceback (most recent call last):
  File "/work/virtual-python-envs/python-38/bin/spacy", line 8, in <module>
    sys.exit(app())
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/typer/main.py", line 214, in __call__
    return get_command(self)(*args, **kwargs)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/typer/main.py", line 497, in wrapper
    return callback(**use_params)  # type: ignore
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/spacy/cli/train.py", line 56, in train_cli
    nlp = init_nlp(config, use_gpu=use_gpu)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/spacy/training/initialize.py", line 55, in init_nlp
    nlp.initialize(lambda: train_corpus(nlp), sgd=optimizer)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/spacy/language.py", line 1232, in initialize
    proc.initialize(get_examples, nlp=self, **p_settings)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/spacy/pipeline/textcat.py", line 365, in initialize
    self.model.initialize(X=doc_sample, Y=label_sample)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/concatenate.py", line 129, in init
    layer.initialize(X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 294, in initialize
    validate_fwd_input_output(self.name, self._func, X, Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/util.py", line 424, in validate_fwd_input_output
    raise DataValidationError(name, X, Y, e.errors()) from None
thinc.util.DataValidationError:

Data validation error in 'para-attn'
X: <class 'thinc.types.Ragged'> Y: <class 'numpy.ndarray'>

Y -> 0   instance of Ragged expected
`

**Comment by svlandeg** [comment]
Great, thanks both! I'll look into it.

**Comment by svlandeg** [comment]
I found the culprit but I'm not sure yet about the best way to fix this. Thanks for your patience :-)

In the meantime, if you want to continue experimenting, you can change `spacy.TextCatEnsemble.v2` to `custom_TextCatEnsemble` and register that function as such:

```
@registry.architectures.register("custom_TextCatEnsemble")
def build_custom_text_classifier(
    tok2vec: Model[List[Doc], List[Floats2d]],
    linear_model: Model[List[Doc], Floats2d],
    nO: Optional[int] = None,
) -> Model[List[Doc], Floats2d]:
    exclusive_classes = not linear_model.attrs["multi_label"]
    with Model.define_operators({">>": chain, "|": concatenate}):
        width = tok2vec.maybe_get_dim("nO")
        print("width", width)
        cnn_model = (
                tok2vec
                >> list2ragged()
                >> reduce_sum()
                >> residual(Maxout(nO=width, nI=width))
                >> Linear(nO=nO, nI=width)
                >> Dropout(0.0)
        )

        nO_double = nO * 2 if nO else None
        if exclusive_classes:
            output_layer = Softmax(nO=nO, nI=nO_double)
        else:
            output_layer = Linear(nO=nO, nI=nO_double) >> Dropout(0.0) >> Logistic()
        model = (linear_model | cnn_model) >> output_layer
        model.set_ref("tok2vec", tok2vec)
    if model.has_dim("nO") is not False:
        model.set_dim("nO", nO)
    model.set_ref("output_layer", linear_model.get_ref("output_layer"))
    model.attrs["multi_label"] = not exclusive_classes
    return model

**Comment by thomashacker** [comment]
Thank you for your post!
I registered the custom textcat model and changed it in the config.

Unfortunately i received this error when training:

```
=========================== Initializing pipeline ===========================
width None
Set up nlp object from config
Pipeline: ['transformer', 'textcat']
Created vocabulary
Finished initializing nlp object
Traceback (most recent call last):
  File "/work/virtual-python-envs/python-38/bin/spacy", line 8, in <module>
    sys.exit(app())
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/typer/main.py", line 214, in __call__
    return get_command(self)(*args, **kwargs)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/typer/main.py", line 497, in wrapper
    return callback(**use_params)  # type: ignore
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/spacy/cli/train.py", line 56, in train_cli
    nlp = init_nlp(config, use_gpu=use_gpu)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/spacy/training/initialize.py", line 55, in init_nlp
    nlp.initialize(lambda: train_corpus(nlp), sgd=optimizer)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/spacy/language.py", line 1232, in initialize
    proc.initialize(get_examples, nlp=self, **p_settings)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/spacy/pipeline/textcat.py", line 365, in initialize
    self.model.initialize(X=doc_sample, Y=label_sample)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/concatenate.py", line 129, in init
    layer.initialize(X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 296, in initialize
    self.init(self, X=X, Y=Y)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/chain.py", line 90, in init
    curr_input = layer.predict(curr_input)
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/model.py", line 312, in predict
    return self._func(self, X, is_train=False)[0]
  File "/work/virtual-python-envs/python-38/lib/python3.8/site-packages/thinc/layers/residual.py", line 49, in forward
    return X + Y, backprop
ValueError: operands could not be broadcast together with shapes (10,768) (10,3)

```

**Comment by svlandeg** [comment]
Yea, I just noticed (and fixed locally) that as well. The problem was a bit more fundamental than I thought, and it's going to require an update in spaCy, spacy-transformers and Thinc. Bear with me ;-)

**Comment by svlandeg** [comment]
One more thing you could try in the meantime is to replace the textcat architecture with `spacy.TextCatCNN.v1`. This architecture requires the same `tok2vec` and `nO` parameters, and `exclusive_classes` from the `linear_model`'s `multi_label` config (and then you can remove the `linear_model` altogether)

**Comment by thomashacker** [comment]
This worked thanks!

**Comment by svlandeg** [comment]
Happy to hear it! We're working on a proper fix for the Ensemble architecture, but it'll require a few PR's and new releases, so that may take a little while: PR https://github.com/explosion/spaCy/pull/6431

**Comment by lbenc135** [comment]
Hey @svlandeg any update on this? Let me know if there's any help needed. Btw is there any estimate for when rc3 is coming?

**Comment by svlandeg** [comment]
This issue should be resolved by the PRs I linked above, and they will be available in the next rc3 release. The relevant PR on `spacy-transformers` is pending and will be merged once the rc3 is out. And once that is done, we'll release the new version of the transformers library and together the issue should be fixed.

It's always difficult to promise a certain release schedule because unexpected issues may come up. But we're hoping to get the rc3 out next week!

**Comment by svlandeg** [comment]
This should all be fixed in the current release of spaCy 3.0.1, so I'll be closing this. Feel free to open a new issue if you would run into any new trouble with the current release though ;-)
